### PR TITLE
Add print of form specification

### PR DESCRIPTION
**Pull request by LaChope**
Resolves https://github.com/kbss-cvut/23ava-distribution/issues/46
To get the form specifications, simply go in console and run `console.log(FormUtils.getFormSpecification(questions))`

### CODE REVIEW

File: src/util/FormUtils.js
Comment (blcham) [inline src/util/FormUtils.js:489]: I suggest not having in the name "onEnterAndLeave" as it is clear from the parameters and DFS algorithmitself.

IMO, DFS algorithm itself clearly communicates entering and leaving the node ... e.g. by using stack (LIFO) where it holds nodes. It is strange to me to say it explicitly in the method name in the same matter as saying `fetchData()` ->  `fetchToReturnData()` or `deleteQuestion()` --> `deleteToNotHaveQuestion()`.

Before, the algorithm used onEnter question, which is IHO typically expected, but should be said explicitly, now, we just extended the algorithm.

File: src/components/FormManager.jsx
Comment (blcham) [inline src/components/FormManager.jsx:45]: I would rather have a method to return string than print to console ... maybe I suggested the name incorrectly

User can always do console.log(printFormSpecification).

Maybe `getFormSpecification` ?

File: src/util/FormUtils.js
Comment (blcham) [inline src/util/FormUtils.js:489]: I do not see the reason to define new method. Citing myself from the issue description:
"""
Thus I suggest adding `FormUtils.printFormSpecification` that would use modified dfsTraverse function. The modified dfsTraverse function would have 3 parameters, while 3. parameter should be optional, i.e.
`dfsTraverse(startingQuestions:array, onEnterQuestion:function, onLeaveQuestion:function)`
"""

Comment (blcham) [reply src/util/FormUtils.js:489]: It makes sense to me to just modify the existing one while usage does not need to be refactored as the modification is backward compatible as 3rd param is optional (i.e. defaulting to the empty method -- (question) => {}) .

File: src/util/FormUtils.js
Comment (blcham) [inline src/util/FormUtils.js:528]: I prefer to have 3 values of one property as suggested, e.g.:
`requirement: NONE`
`requirement: REQUIRED_TO_SAVE_FORM
`requirement: REQUIRED_TO_COMPLETE_FORM`



My motivation here is that it is easier to read/search and understand for the enduser.

File: .storybook/preview.js
Comment (blcham) [inline .storybook/preview.js:133]: Not sure that we need it?  I need your opinion.

It is really rather functionality targeted to developers so I do not want to expose it to other projects like record-manager-ui. Is this a standard (best practice) way how to solve such an issue? 

I can imagine 2 situations in how to use it:
- for developer --  debugging the form 
- for application that uses the library -- it can show specification the form to the user if form rendering fails. This would have to be think through -- so let's not discuss it here.

So my question is, is it impossible to use in the development console without exposing it to the `window`? I would rather extend development notes within the SForms than pollute the `window`.
Moreover, polluting it with FormUtils is strange in sense that most of the methods in FormUtils are not useful for the developer, or at least as useful as e.g. JsonLdUtils, which we do not expose ...

**Review by blcham — CHANGES_REQUESTED** [review]
see my comments

File: .storybook/preview.js
Comment (LaChope) [reply .storybook/preview.js:133]: Without exposing it to window, we cannot use it "easily" as a script in the console. The alternatives I see are:
- Putting it in a different javascript and run it with something like `node PrintFormSpecification.js`. However, it is hard in javascript to run a specific function from CLI (we would have to add some tooling, according to [this post](https://stackoverflow.com/questions/30782693/run-function-in-script-from-command-line-node-js), the main issue is that we are using a `module`). So we would have to run it as it is as change the questions directly inside the JS file
- Write a bash script

However, to run from the console I do not see any other alternative but to expose it to the window. We could create another JS file and expose only this specific file and function

File: src/util/FormUtils.js
Comment (blcham) [reply src/util/FormUtils.js:489]: @LaChope

File: src/util/FormUtils.js
Comment (blcham) [reply src/util/FormUtils.js:528]: @LaChope

File: src/util/FormUtils.js
Comment (LaChope) [reply src/util/FormUtils.js]: Agree

File: .storybook/preview.js
Comment (LaChope) [reply .storybook/preview.js]: It is now working in the toolbar of storybook: 
image

Comment (blcham) [reply .storybook/preview.js]: Looks great !!! It can also be copied to clipboard (although i have to select it all) so it is very useful.

However, i see weird output there, which i suggest to formulate as new issue while closing this one:
image

File: src/util/FormUtils.js
Comment (blcham) [inline src/util/FormUtils.js:470]: I suggest the following:
```
/**
 * Traverse provided array of questions recursively by DFS algorithm
 * and execute on each traversed question provided function.
 * @param {Array<Object>} questions - The array of questions to be traversed.
 * @param {Function} onEnterQuestion - The callback function to execute when entering a question. 
 * @param {Function} [onLeaveQuestion=()=>{}] - The callback function to execute when leaving a question.
 */
static dfsTraverseQuestionTree(questions, onEnterQuestion, onLeaveQuestion = () => {}) {
  function recursiveTraverse(question) {
    onEnterQuestion(question);
    Utils.asArray(question[Constants.HAS_SUBQUESTION]).forEach((q) => {
      recursiveTraverse(q);
    });
    onLeaveQuestion(question);
  }

  questions.forEach((q) => {
    recursiveTraverse(q);
  });
}
```

**Review by blcham — CHANGES_REQUESTED** [review]
LGTM, after resolution of comments can be merged.

File: .storybook/preview.js
Comment (LaChope) [reply .storybook/preview.js]: Ticket was created here: https://github.com/kbss-cvut/s-forms/issues/310

**Pull request merged**
